README.md
<img src="man/figures/ALG_1_dat-ALG_2_dat.png" width="455" height="242" alt= "EAF
   differences between two bi-objective optimizers" title= "EAF differences
   between two bi-objective optimizers" style= "border:0; align:right; float:right;"/>
<img src="man/figures/eafdiff-color.png" width="480"
   height="240" alt="EAF differences between two variants of W-RoTS (color)"
   title="EAF differences between two variants of W-RoTS (color)"
   style="border:0; align:right; float:right; clear:right"/>
   
**eaf**: Empirical Attainment Function (EAF) Tools
================================================================

[![CRAN Status](https://www.r-pkg.org/badges/version-last-release/eaf)](https://cran.r-project.org/package=eaf) [![CRAN Downloads](https://cranlogs.r-pkg.org/badges/grand-total/eaf)](https://CRAN.R-project.org/package=eaf) [![Travis Build
Status](https://travis-ci.org/MLopez-Ibanez/eaf.svg?branch=master)](https://travis-ci.org/MLopez-Ibanez/eaf)

[ [**Homepage**](http://lopez-ibanez.eu/eaftools) ]  [ [**GitHub**](https://github.com/MLopez-Ibanez/eaf) ]

**Maintainer:** [Manuel López-Ibáñez](http://lopez-ibanez.eu)

**Contributors:**
    [Manuel López-Ibáñez](http://lopez-ibanez.eu),
    [Marco Chiarandini](http://www.imada.sdu.dk/~marco),
    [Carlos M. Fonseca](http://eden.dei.uc.pt/~cmfonsec),
    [Luís Paquete](http://eden.dei.uc.pt/~paquete),
    [Thomas Stützle](http://iridia.ulb.ac.be/~stuetzle),
    and Mickaël Binois.
    
---------------------------------------

Introduction
============

The empirical attainment function (EAF) describes the probabilistic
distribution of the outcomes obtained by a stochastic algorithm in the
objective space. This [R](http://www.r-project.org/) package implements plots
of summary attainment surfaces and differences between the first-order
EAFs. These plots may be used for exploring the performance of stochastic local
search algorithms for biobjective optimization problems and help in identifying
certain algorithmic behaviors in a graphical way.

The corresponding [book chapter](#LopPaqStu09emaa) [1] explains the use of
these visualization tools and illustrates them with examples arising from
practice.

**Keywords**: empirical attainment function, summary attainment surfaces, EAF
differences, multi-objective optimization, bi-objective optimization,
performance measures, performance assessment, graphical analysis,
visualization.

**Relevant literature:**

 1. <a name="#LopPaqStu09emaa"/>Manuel López-Ibáñez, Luís Paquete, and Thomas Stützle. [Exploratory Analysis of Stochastic Local Search Algorithms in Biobjective Optimization](http://dx.doi.org/10.1007/978-3-642-02538-9_9). In T. Bartz-Beielstein, M. Chiarandini, L. Paquete, and M. Preuss, editors, *Experimental Methods for the Analysis of Optimization Algorithms*, pages 209–222. Springer, Berlin, Germany, 2010.<br>
    (This chapter is also available in a slightly extended form as [Technical Report TR/IRIDIA/2009-015](http://iridia.ulb.ac.be/IridiaTrSeries/IridiaTr2009-015r001.pdf)).<br>
    [ [bibtex](http://lopez-ibanez.eu/LopezIbanez_bib.html#LopPaqStu09emaa) |
    doi: [10.1007/978-3-642-02538-9_9](http://dx.doi.org/10.1007/978-3-642-02538-9_9)
    | [Presentation](http://lopez-ibanez.eu/doc/gecco2010moworkshop.pdf) ]


Download and installation
-------------------------

The software is implemented as an R package, but the code for only computing the EAF is available as a C program, and it does not require installing R or any R packages. Just [download the package source code](https://cran.r-project.org/package=eaf), uncompress it, and look for the directory `inst/scripts/eaf`. This code can be used to implement your own visualizations instead of the visualizations provided by the eaf package.

The visualization of the EAFs require installing the eaf package. Therefore, for making use of all the features, a basic knowledge of R is recommended. However, the eaf package contains two Perl scripts that allow to generate standard plots without any knowledge about R. See `inst/scripts/eafplot/` and `inst/scripts/eafdiff/` in the package source code. The scripts use the eaf package internally to generate the plots, and, hence, the eaf package must be installed and working.

The first step before installing the eaf package is to [install R](https://cran.r-project.org/). Once R is installed in the system, there are two methods for installing the eaf package:

 1. Install within R (automatic download, internet connection required). Invoke
    R, then
    
    ```r
        install.packages(c("modeltools", "eaf"))
    ```
    
 2. (Download the eaf package from CRAN)[https://cran.r-project.org/package=eaf] (you may also need to download and install
    first the (package modeltools)[https://cran.r-project.org/package=modeltools]), and invoke at the command-line:
    
    ```bash
        R CMD INSTALL <package>
    ```
    where `<package>` is one of the three versions available: `.tar.gz` (Unix/BSD/GNU/Linux), `.tgz` (MacOS X), or `.zip` (Windows).

Search the [R documentation](https://cran.r-project.org/faqs.html) if you need more help to install an R package on your system.

If you wish to be notified of bugfixes and new versions, please subscribe to the [low-volume emo-list](https://lists.dei.uc.pt/mailman/listinfo/emo-list), where announcements will be made.

[ [Download eaf package from CRAN](https://cran.r-project.org/package=eaf) ]  [ [Documentation](http://lopez-ibanez.eu/eaftools) ]  [ [Development version (GitHub)](https://github.com/MLopez-Ibanez/eaf) ]


Usage
-----

Once the eaf package is installed, the following R commands will give more information:
```r
    library(eaf)
    ?eaf
    ?eafplot
    ?eafdiffplot
    ?read.data.sets
    example(eafplot)
    example(eafdiffplot) # This one takes some time
```

Apart from the main R package, the source code contains the following extras in
the directory `inst/` (after installation, these files can be found at the
directory printed by the R command `system.file(package="eaf")`):

 * `scripts/eaf` : This C program computes the empirical attainment function in 2 dimensions. It is NOT required by the other programs, but it is provided as a useful command-line utility. This version is based on the original code written by Carlos M. Fonseca available at http://www.tik.ee.ethz.ch/pisa/. Another version of the code that handles three dimensions is available at [Prof. Fonseca's website](http://eden.dei.uc.pt/~cmfonsec/software.html#aft).
 * `scripts/eafplot` : Perl script to plot summary attainment surfaces.
 * `scripts/eafdiff` : Perl script to plot the differences between the EAFs of two input sets.
 * `extdata/` : Examples of utilization of the above programs. These are discussed in the corresponding [book chapter](#LopPaqStu09emaa) [1].

For more information, consult the `README` files at each subdirectory.


Python
------

Thanks to [rpy2](https://rpy2.github.io/doc/latest/html/index.html), you can
use the `eaf` package from Python. A complete example would be:

```python
    import numpy as np
    from rpy2.robjects.packages import importr
    from rpy2.robjects import r as R
    from rpy2.interactive import process_revents
    process_revents.start()

    eaf = importr("eaf")
    path = R('system.file(package="eaf")')[0] + "/extdata/"
    alg1 = eaf.read_data_sets_(path + "ALG_1_dat")
    alg1 = np.array(alg1).transpose()
    eaf.eafplot(alg1[:, 0:2], sets=alg1[:,2])

    alg2 = np.array(eaf.read_data_sets_(path + "ALG_2_dat")).transpose()
    eaf.eafdiffplot(alg1, alg2)
```

License
--------

This software is Copyright (C) 2011 Carlos M. Fonseca, Luís Paquete, Thomas
Stützle, Manuel López-Ibáñez and Marco Chiarandini.

This program is free software (software libre); you can redistribute it and/or
modify it under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the [GNU General Public License](http://www.gnu.org/licenses/gpl.html) for more details.

**IMPORTANT NOTE**: Please be aware that the fact that this program is released
as Free Software does not excuse you from scientific propriety, which obligates
you to give appropriate credit! If you write a scientific paper describing
research that made substantive use of this program, it is your obligation as a
scientist to (a) mention the fashion in which this software was used in the
Methods section; (b) mention the algorithm in the References section. The
appropriate citation is:

 * Manuel López-Ibáñez, Luís Paquete, and Thomas Stützle. **Exploratory Analysis of Stochastic Local Search Algorithms in Biobjective Optimization.** In T. Bartz-Beielstein, M. Chiarandini, L. Paquete, and M. Preuss, editors, *Experimental Methods for the Analysis of Optimization Algorithms*, pages 209–222. Springer, Berlin, Germany, 2010.  doi: 10.1007/978-3-642-02538-9_9

Moreover, as a personal note, I would appreciate it if you would email
`manuel.lopez-ibanez@manchester.ac.uk` with citations of papers referencing
this work so I can mention them to my funding agent and tenure committee.


NEWS
* Compute Vorob'ev threshold, expectation and deviation. Plots of the symmetric
  deviation.
                                                              (Mickael Binois)
* Non-integer percentiles are computed correctly.
                                                         (Manuel López-Ibáñez)

* Various aesthetic improvements in eafplot and eafdiffplot.
                                                         (Manuel López-Ibáñez)

* Functions for computing hypervolume, hypervolume contributions, epsilon
  metric, IGD+, filter dominated points, and fast normalisation of ranges.
                                                         (Manuel López-Ibáñez)


1.8

* Development version moved to GitHub: https://github.com/MLopez-Ibanez/eaf

* Remove leading zeros from version number.

* New parameters left.panel.last and right.panel.last of eafdiffplot.

* Export and document function 'eafs' to compute EAFs.

* eafdiff.pl: Handle --colors=, --intervals=.
 
* Fix crash in eafplot if 'sets' is a vector of strings.

* Reset layout in eafdiffplot.

* Compute eaf for 3D.

* Added testthat testing framework.

* The documentation is now generated with Roxygen2.

* Entry points to C code are now properly registered.


1.07

* Silence CRAN warning for GNU extensions in Makefiles in
  inst/scripts/eaf/Makefile.


1.06

* Fix bug when automatically generating a legend in eafplot with the
  formula interface (Thanks to Bernd Bischl for reporting this)

* Improve handling of various newline character formats.

* Reduce memory consumption (up to four times less memory).

* eafdiff.pl: Mention option --legendpos=none to hide the legend.

* eafplot.pl: Add options --maximise and --xmaximise and --colors=.
  Fix bug with --area.

* eafplot.default now requires two colors when type=="area" and a
  palette is interpolated between these two colors for all the
  different levels plotted (Thanks to Alexandre Quemy for the suggestion).


1.05

* Implement eafdiffplot(, type = "area"). This is now the default. 

  The idea for the algorithm to compute the areas was provided by
  Carlos M. Fonseca.  The implementation uses R polygons, which some
  PDF viewers may have trouble rendering correctly (See
  https://cran.r-project.org/doc/FAQ/R-FAQ.html#Why-are-there-unwanted-borders).
  Plots will look correct when printed. To get the previous behaviour
  use eafdiffplot(, type = "point").
  
* Attempt to deal with DOS/Unix newlines correctly.

* eafplot.pl: Lots of cleanup.
  (--single, --output, --legend): New options.
  Default to PDF.

* eafdiff.pl: Many cleanups.
  EAF diff with type="area" is the default now.
  Parameter --cex replaced by --scale.
  Do not require ps2eps.
  Handle --obj1= better. Crop pdf before converting to png.
  (--noattsurfs): New option.
  (--output-dir,--output): New options.
  (--eps): New option. PDF is the default now.

* Force eafdiff plots to use a square plotting region.

* read.data.sets() normalizes paths, thus it works with files such as ~/file.

* Function eafdiffplot() handles percentiles=NA and grand.lines
  parameter.

* Fix bug with ranges and maximise=TRUE in eafdiffplot().

* Fix points.steps to work correctly for all values of maximise.

* Fix bug in  eafdiffplot(..., maximise=c(TRUE,FALSE), full.eaf = TRUE)

* Fix "log" parameter in eafplot and eafdiffplot.

* Avoid that eafplot.formula modifies global options.

* Add 'axes' parameter to eafplot.

* To avoid confusion, the arguments 'xaxis.side' and 'yaxis.side' of
  eafplot can only take values c("below", "above") and c("left",
  "right"), respectively.

* Add missing Makefile in inst/scripts/eaf/.

1.04

* Fix issues with only one point in the EAF.

1.03

* Handle maximise argument in eafdiffplot and --maximise command-line
  option in eafdiff.pl

* R/calls.R (eafplot.data.frame): Fix problem with main parameter.

1.00 

* Initial release available in CRAN.


R/calls.R
###############################################################################
#
#                          Copyright (c) 2011-2018
#         Manuel Lopez-Ibanez <manuel.lopez-ibanez@manchester.ac.uk>
#             Marco Chiarandini <marco@imada.sdu.dk>
#
# This program is free software (software libre); you can redistribute
# it and/or modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, you can obtain a copy of the GNU
# General Public License at http://www.gnu.org/copyleft/gpl.html
#
# IMPORTANT NOTE: Please be aware that the fact that this program is
# released as Free Software does not excuse you from scientific
# propriety, which obligates you to give appropriate credit! If you
# write a scientific paper describing research that made substantive use
# of this program, it is your obligation as a scientist to (a) mention
# the fashion in which this software was used in the Methods section;
# (b) mention the algorithm in the References section. The appropriate
# citation is:
# 
#    Manuel Lopez-Ibanez, Luis Paquete, and Thomas Stuetzle.
#    Exploratory Analysis of Stochastic Local Search Algorithms in
#    Biobjective Optimization. In T. Bartz-Beielstein, M. Chiarandini,
#    L. Paquete, and M. Preuss, editors, Experimental Methods for the
#    Analysis of Optimization Algorithms, pages 209-222. Springer,
#    Berlin, Germany, 2010. doi: 10.1007/978-3-642-02538-9_9
# 
# Moreover, as a personal note, I would appreciate it if you would email
# manuel.lopez-ibanez@manchester.ac.uk with citations of papers referencing this
# work so I can mention them to my funding agent and tenure committee.
#
################################################################################
#
# TODO:
#
#  * Follow this style for coding:
#    http://google-styleguide.googlecode.com/svn/trunk/google-r-style.html
#
################################################################################

#' Plot the Empirical Attainment Function for two objectives generic
#' 
#' Computes and plots the Empirical Attainment Function, either as
#' attainment surfaces for certain percentiles or as points.
#'
#' This function can be used to plot random sets of points like those obtained
#' by different runs of biobjective stochastic optimization algorithms.  An EAF
#' curve represents the boundary separating points that are known to be
#' attainable (that is, dominated in Pareto sense) in at least a fraction
#' (quantile) of the runs from those that are not. The median EAF represents
#' the curve where the fraction of attainable points is 50\%.  In single
#' objective optimization the function can be used to plot the profile of
#' solution quality over time of a collection of runs of a stochastic optimizer.
#' 
#' @param x Either a matrix of data values, or a data frame, or a list of
#'     data frames of exactly three columns.
#'
#' @keywords graphs
#' @export
eafplot <- function(x, ...) UseMethod("eafplot")


#' @describeIn eafplot Formula interface
#'
#'@param formula A formula of the type: \code{time + cost ~ run | instance}
#'     will draw \code{time} on the x-axis and \code{cost} on the y-axis. If \code{instance} is
#'     present the plot is conditional to the instances.
#' 
#'@param data Dataframe containing the fields mentioned in the formula and in groups.
#'@export
eafplot.formula <- function(formula, data, groups = NULL, subset = NULL, ...)
{
  ## formula of type time+cost~run|inst, groups=alg
  if (missing(formula))
    stop("formula missing")
 
  if ((length(formula) != 3L) || (length(formula[[2L]]) != 3L))
    stop("incorrect specification for 'formula'")

  mf <- modeltools::ModelEnvFormula(formula = formula, data = data,
                                    subset = subset, designMatrix = FALSE,
                                    responseMatrix = FALSE, ...)

  ### extract data from the ModelEnv object
  points <- mf@get("response")
  sets <- mf@get("input")
  cond <- NULL
  if (length(mf@formula) == 3L)
    cond <- as.list(mf@get("blocks"))

  groups <- eval(substitute(groups), data, environment(formula))

  if (!is.null(groups) && !is.factor(groups))
    stop("groups must be a factor")
  
  if (length(cond) == 0)
    {
      strip <- FALSE
      cond <- list(gl(1, length(points)))
    }

  condlevels <- lapply(cond, levels)
  cond.max.level <- unlist(lapply(cond, nlevels))

  npackets <- prod(cond.max.level)

  panel.args <- vector(mode = "list", length = npackets)

  packet.sizes <- numeric(npackets)
  if (npackets > 1)
    {
      dim(packet.sizes) <- sapply(condlevels, length)
      dimnames(packet.sizes) <- lapply(condlevels, as.character)
    }

  cond.current.level <- rep(1, length(cond))

  for (packet.number in seq_len(npackets)) {
    id <- .compute.packet(cond, cond.current.level)
    packet.sizes[packet.number] <- sum(id)

    panel.args[[packet.number]] <-
      list(points = as.matrix(points[id,]), sets = as.numeric(sets[id,]),
           groups=groups[id])
    ## MARCO: I do not think we need to care about subscripts... or do we?
    #if (subscripts)
    #   panel.args[[packet.number]]$subscripts <-
    #                   subscr[id]
    cond.current.level <- .cupdate(cond.current.level, cond.max.level)
  }
  # FIXME: I don't think this is doing the right thing.
  op <- par(mfrow = .check.layout(NULL,cond.max.level)[2:3])
  on.exit(par(op))
  for (i in seq_len(length(panel.args))) {
    eafplot.default(panel.args[[i]]$points,
                    panel.args[[i]]$sets,
                    panel.args[[i]]$groups,
                    ...)
  }
  invisible()
}


#' @describeIn eafplot List interface for lists of data.frames
#' 
#'@export 
eafplot.list <- function(x,...)
{
  if (!is.list(x))
    stop("'x' must be a list of data.frames with exactly three columns")

  DT <- data.frame()
  if (!is.null(names(x)))
    groups <- names(x)
  else
    groups <- 1:length(x)
  for (i in seq_len(length(x)) ) {
    if (!is.data.frame(x[[i]]))
      stop("Each element of the list must be a data.frame with exactly three columns.")
    DT <- rbind(DT, data.frame(x[[i]], groups = groups[i]))
  }
  eafplot(as.matrix(DT[,c(1,2)]), as.numeric(as.factor(DT[,3])), as.factor(DT[,4]),...)
}

#' @describeIn eafplot Data.frame interface
#'
#' @param y Either a matrix of data values, or a data frame.
#' @export
eafplot.data.frame <- function(x, y = NULL, sets = NULL, ...)
{
  namex <- deparse(substitute(x))
  namey <- deparse(substitute(y))

  # FIXME: Why keep x and y as data.frame to convert it here to matrix?
  eafplot.data.frame2 <- function(x, groups, main = DNAME, ...)
    eafplot(as.matrix(x[,c(1,2)]), as.numeric(x[,3]),
            groups = groups, main = main, ...)
  check.eaf.data.frame <- function(x, sets = NULL) {
    xname <- deparse(substitute(x))
    if (!is.null(sets))
      x <- cbind.data.frame(x, sets)

    if (!is.data.frame(x) || ncol(x) != 3L)
      stop("'", xname, "' must be a data.frame with exactly three columns.\n",
           "  If you have grouping and conditioning variables, please consider using this format: 'eafplot(formula, data, ...)'")
    if (nrow(x) < 1L)
      stop("not enough (finite) '", xname, "' observations")
    if (!is.numeric(x[,1]) || !is.numeric(x[,2]))
      stop("columns 1 and 2 of '", xname, "' must be numeric")
    if (!is.numeric(x[,3]) && !is.factor(x[,3]))
      x[,3] <- as.factor(x[,3])
    return(x)
  }
  x <- check.eaf.data.frame(x, sets)
  if (!is.null(y)) {
    y <- check.eaf.data.frame(y, sets)
    DNAME <- paste0(namex, " and ", namey)
    DT <- rbind(data.frame(x, groups = namex),
                data.frame(y, groups = namey))
    groups <- as.factor(DT[, 4])
  } else {
    DNAME <- namex
    DT <- x
    groups <- NULL
  }
  eafplot.data.frame2(DT, groups = groups, ...)
}


### Local Variables:
### ess-style: DEFAULT
### indent-tabs-mode: nil
### ess-fancy-comments: nil
### End:

